core/com.b2international.snowowl.datastore.server/src/com/b2international/snowowl/datastore/server/internal/CDOBasedRepository.java

```diff
@@ -348,5 +363,81 @@ public void handleCommitInfo(CDOCommitInfo commitInfo) {
 			notification.publish(events());
         }
 	}
-	
+
+	@Override
+	public boolean isConsistent(IBranchPath branch) {
+		List<CDOCommitInfo> cdoCommitInfos = getCDOCommitInfos(branch);
+		CommitInfos indexCommitInfos = getIndexCommitInfos(branch);
+		
+		boolean emptyDb = cdoCommitInfos.isEmpty();
+		boolean emptyIndex = indexCommitInfos.getItems().isEmpty();
+
+		if (emptyDb && emptyIndex) {
+			return true; // empty dataset
+		}
+		
+		if (emptyDb ^ emptyIndex) {
+			LOG.error("{} is in inconsistent state. CDO {} but INDEX {}", getCdoRepository().getRepositoryName(), contentMessage(emptyDb), contentMessage(emptyIndex));
+			return false; // either CDO or index was deleted but not the other.
+		}
+		
+		return !hasInvalidCDOTimeStamps(cdoCommitInfos, getLast(indexCommitInfos).getTimeStamp(), branch);
```

Comment:
To be absolutely, positively sure that we have the same content, commit infos coming from CDO and the index should be compared pairwise. The two lists should have an equal length, and the timestamps should match when iterating over the two lists in lockstep (the lists should be sorted by commit timestamp first if they are not in this form already). 

Such a comparison would also cover the checks above (non-zero CDO commit infos against zero indexed commit infos, or the other way round).